### PR TITLE
GPU GridLayer PART-3: Add NewGridLayer.

### PR DESCRIPTION
**Pull request by 1chandu**
For #2084  

#### Background

- Adding `NewGridLayer`, a composite layer that does some prop conversion and either renders `GridLayer` (cpu aggregation) or `GPUGridLayer`.
    - After all testing, `GridLayer` becomes `CPUGridLayer` and `NewGridLayer` becomes `GridLayer`.
- GPUGridCellLayer
    - Fix lighting.
    - Match elevation calculations with `ColumnLayer`.
- Tested `NewGridLayer`'s CPU and GPU paths with existing `GridLayer` golden image and in `layer-browser` example.
- Testing and docs will be added in future PRs , tracked in #2084 

#### Change List
- Add NewGridLayer, fixes to `gpu-grid-cell-layer`.

### CODE REVIEW

File: modules/aggregation-layers/src/grid-layer/grid-layer.js [hunk 1]
```diff
@@ -349,3 +351,4 @@ export default class GridLayer extends CompositeLayer {
 
 GridLayer.layerName = 'GridLayer';
 GridLayer.defaultProps = defaultProps;
+GridLayer.DEFAULT_GETVALUE = DEFAULT_GETVALUE;
```
Comment (Pessimistress) [inline modules/aggregation-layers/src/grid-layer/grid-layer.js]: Why is this needed?

File: modules/aggregation-layers/src/new-grid-layer/new-grid-layer.js [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+import GPUGridLayer from '../gpu-grid-layer/gpu-grid-layer';
+import GridLayer from '../grid-layer/grid-layer';
+
+const defaultProps = Object.assign({}, GPUGridLayer.defaultProps, GridLayer.defaultProps);
+
+// Function to convert from getWeight/aggregation props to getValue prop for Color and Elevation
+function getMean(pts, accessor) {
+  const filtered = pts.map(item => accessor(item)).filter(pt => Number.isFinite(pt));
+
+  return filtered.length ? filtered.reduce((accu, curr) => accu + curr, 0) / filtered.length : null;
```
Comment (Pessimistress) [inline modules/aggregation-layers/src/new-grid-layer/new-grid-layer.js]: Perf: avoid defining inline function in frequently used utility functions.

```js
const filtered = pts.map(accessor).filter(Number);
filtered.reduce(sumReducer, 0);
```

File: modules/aggregation-layers/src/new-grid-layer/new-grid-layer.js [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    this.setState(newState);
+  }
+
+  renderLayers() {
+    let gridLayer = null;
+
+    if (this.state.useGPUAggregation) {
+      gridLayer = new GPUGridLayer(
+        this.props,
+        this.getSubLayerProps({
```
Comment (Pessimistress) [inline modules/aggregation-layers/src/new-grid-layer/new-grid-layer.js]: Lacking support for update triggers?

File: modules/aggregation-layers/src/new-grid-layer/new-grid-layer.js [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        this.props,
+        buildColorElevationProps,
+        this.getSubLayerProps({
+          id: 'CPU',
+          // Note: data has to be passed explicitly like this to avoid being empty
+          data: this.props.data
+        })
+      );
+    }
+    return [gridLayer];
```
Comment (Pessimistress) [inline modules/aggregation-layers/src/new-grid-layer/new-grid-layer.js]: You can return a single layer without an array.

**Comment by Pessimistress** [comment]
Add a basic unit test like [this](https://github.com/uber/deck.gl/blob/master/test/modules/layers/polygon-layer.spec.js)

File: modules/aggregation-layers/src/grid-layer/grid-layer.js (hunk 1 above)
Comment (1chandu) [reply modules/aggregation-layers/src/grid-layer/grid-layer.js]: `getColorValue` and `getElevationValue` are function props in the current API. When user provides values for these, I can not map them to new API (like, `getColorValue` to `getColorWeight` and `colorAggregation`), hence I have to fallback to CPU Aggregation. By defining a constant, `DEFAULT_GETVALUE` for default value, I can check if the user provided a custom prop or we are using a default value. For the case of default value, (which is `points => points.length`) I am converting them to new props.

Comment (Pessimistress) [reply modules/aggregation-layers/src/grid-layer/grid-layer.js]: Maybe we can just use `GridLayer.defaultProps. getColorValue.value`?

File: modules/aggregation-layers/src/new-grid-layer/new-grid-layer.js (hunk 3 above)
Comment (1chandu) [reply modules/aggregation-layers/src/new-grid-layer/new-grid-layer.js]: Nothing happening in this layer, no attribute calculations or aggregation. It basically checking if aggregation can happen on GPU or not. `GridLayer` and `GPUGridLayer` have update triggers setup.

Comment (Pessimistress) [reply modules/aggregation-layers/src/new-grid-layer/new-grid-layer.js]: You still need to pass the update triggers to the sublayers. Here the update triggers in `this.props` is overwritten by `getSubLayerProps`.

File: modules/aggregation-layers/src/grid-layer/grid-layer.js (hunk 1 above)
Comment (1chandu) [reply modules/aggregation-layers/src/grid-layer/grid-layer.js]: Yeah, that works, updated.

File: modules/aggregation-layers/src/new-grid-layer/new-grid-layer.js (hunk 2 above)
Comment (1chandu) [reply modules/aggregation-layers/src/new-grid-layer/new-grid-layer.js]: Addressed.

File: modules/aggregation-layers/src/new-grid-layer/new-grid-layer.js (hunk 4 above)
Comment (1chandu) [reply modules/aggregation-layers/src/new-grid-layer/new-grid-layer.js]: Done.

File: modules/aggregation-layers/src/new-grid-layer/new-grid-layer.js (hunk 3 above)
Comment (1chandu) [reply modules/aggregation-layers/src/new-grid-layer/new-grid-layer.js]: Done, also update the document page for `getSublayerProps` , removed `Optional` for `updateTriggers` argument. Seems like we always need to pass this to avoid overriding.

File: modules/aggregation-layers/src/gpu-grid-layer/gpu-grid-cell-layer-vertex.glsl.js [hunk 5]
```diff
@@ -89,45 +87,44 @@ float linearScale(vec2 domain, vec2 range, float value) {
...
-
   vec2 colorDomain = vec2(colorData.maxMinCount.a, colorData.maxMinCount.r);
   vec4 color = quantizeScale(colorDomain, colorRange, colors.r);
 
-  // TODO: discard when noRender is true
-  float finalCellSize = noRender ? 0.0 : project_size(cellSize);
-
   float elevation = 0.0;
 
   if (extruded > 0.5) {
```
Comment (Pessimistress) [inline modules/aggregation-layers/src/gpu-grid-layer/gpu-grid-cell-layer-vertex.glsl.js]: Use `uniform bool extruded`

File: modules/aggregation-layers/src/gpu-grid-layer/gpu-grid-layer.js [hunk 6]
```diff
@@ -187,7 +187,7 @@ export default class GPUGridLayer extends CompositeLayer {
 
     // return props to the sublayer constructor
     return super.getSubLayerProps({
-      id: 'grid-cell',
+      id: 'gpu-grid-cell',
```
Comment (Pessimistress) [inline modules/aggregation-layers/src/gpu-grid-layer/gpu-grid-layer.js]: Instead of overriding the `getSubLayerProps()` method, pass the props directly to `SubLayerClass`:

```js
return new SubLayerClass(
  defaultSubLayerProps,
  this.getSubLayerProps({id, updateTriggers}),
  nonOverridableProps
);
```

It is 1) more performant and 2) allows for user override with the [_subLayerProps prop](https://deck.gl/#/documentation/deckgl-api-reference/layers/composite-layer?section=_sublayerprops-object-experimental).

See https://github.com/uber/deck.gl/blob/master/modules/aggregation-layers/src/hexagon-layer/hexagon-layer.js#L356

File: modules/aggregation-layers/src/gpu-grid-layer/gpu-grid-cell-layer-vertex.glsl.js (hunk 5 above)
Comment (1chandu) [reply modules/aggregation-layers/src/gpu-grid-layer/gpu-grid-cell-layer-vertex.glsl.js]: updated.

File: modules/aggregation-layers/src/gpu-grid-layer/gpu-grid-layer.js (hunk 6 above)
Comment (1chandu) [reply modules/aggregation-layers/src/gpu-grid-layer/gpu-grid-layer.js]: updated.
